test/sparse/higherorderfns.jl

```diff
@@ -0,0 +1,256 @@
+            # --> test broadcast entry point
+            @test broadcast(+, X, Y) == sparse(broadcast(+, fX, fY))
+            @test broadcast(*, X, Y) == sparse(broadcast(*, fX, fY))
+            @test broadcast(f, X, Y) == sparse(broadcast(f, fX, fY))
+            try
+                Base.Broadcast.broadcast_indices(spzeros((shapeX .- 1)...), Y)
+            catch
+                @test_throws DimensionMismatch broadcast(+, spzeros((shapeX .- 1)...), Y)
+            end
+            # --> test broadcast! entry point / +-like zero-preserving op
+            fZ = broadcast(+, fX, fY); Z = sparse(fZ)
+            broadcast!(+, Z, X, Y); Z = sparse(fZ) # warmup for @allocated
+            @test (@allocated broadcast!(+, Z, X, Y)) == 0
+            @test broadcast!(+, Z, X, Y) == sparse(broadcast!(+, fZ, fX, fY))
+            # --> test broadcast! entry point / *-like zero-preserving op
+            fZ = broadcast(*, fX, fY); Z = sparse(fZ)
+            broadcast!(*, Z, X, Y); Z = sparse(fZ) # warmup for @allocated
+            @test (@allocated broadcast!(*, Z, X, Y)) == 0
+            @test broadcast!(*, Z, X, Y) == sparse(broadcast!(*, fZ, fX, fY))
+            # --> test broadcast! entry point / not zero-preserving op
+            fZ = broadcast(f, fX, fY); Z = sparse(fZ)
+            broadcast!(f, Z, X, Y); Z = sparse(fZ) # warmup for @allocated
+            @test (@allocated broadcast!(f, Z, X, Y)) == 0
+            @test broadcast!(f, Z, X, Y) == sparse(broadcast!(f, fZ, fX, fY))
+            # --> test shape checks for both broadcast and broadcast! entry points
+            try
+                Base.Broadcast.check_broadcast_indices(indices(Z), spzeros((shapeX .- 1)...), Y)
+            catch
+                @test_throws DimensionMismatch broadcast!(f, Z, spzeros((shapeX .- 1)...), Y)
+            end
+        end
+    end
+end
+
+@testset "broadcast[!] implementation capable of handling >2 (input) sparse vectors/matrices" begin
+    N, M, p = 10, 12, 0.3
+    f(x, y, z) = x + y + z + 1
+    mats = (sprand(N, M, p), sprand(N, 1, p), sprand(1, M, p), sprand(1, 1, 1.0), spzeros(1, 1))
+    vecs = (sprand(N, p), sprand(1, 1.0), spzeros(1))
+    tens = (mats..., vecs...)
+    for Xo in tens
+        X = ndims(Xo) == 1 ? SparseVector{Float32,Int32}(Xo) : SparseMatrixCSC{Float32,Int32}(Xo)
+        # use different types to check internal type stability via allocation tests below
+        shapeX, fX = size(X), Array(X)
+        for Y in tens, Z in tens
+            fY, fZ = Array(Y), Array(Z)
+            # --> test broadcast entry point
+            @test broadcast(+, X, Y, Z) == sparse(broadcast(+, fX, fY, fZ))
+            @test broadcast(*, X, Y, Z) == sparse(broadcast(*, fX, fY, fZ))
+            @test broadcast(f, X, Y, Z) == sparse(broadcast(f, fX, fY, fZ))
+            try Base.Broadcast.broadcast_indices(spzeros((shapeX .- 1)...), Y, Z)
+            catch @test_throws DimensionMismatch broadcast(+, spzeros((shapeX .- 1)...), Y, Z) end
```

Comment:
though I still think it would be preferable to have this serve as an independent verification of how `broadcast_indices` is expected to behave for the differently-sized inputs, rather than relying on it throwing - if it were to be refactored to no longer throw, then this test would silently stop doing anything